src/main/java/frc/robot/utils/Tracer.java

```diff
@@ -0,0 +1,211 @@
+      return str;
+    }
+
+    private String popTraceStack() {
+      traceStack.remove(traceStack.size() - 1);
+      return traceStackHistory.remove(traceStackHistory.size() - 1);
+    }
+
+    private double totalGCTime() {
+      double gcTime = 0;
+      for (GarbageCollectorMXBean gc : gcs) {
+        gcTime += gc.getCollectionTime();
+      }
+      return gcTime;
+    }
+
+    private void endCycle() {
+      // update times for all already existing entries
+      for (var entry : entryArray) {
+        // if the entry isn't found, time will null-cast to 0.0
+        Double time = traceTimes.remove(entry);
+        if (time == null) time = 0.0;
+        Logger.recordOutput("Tracer/" + entry, time);
+      }
+      // log all new entries
+      for (var traceTime : traceTimes.entrySet()) {
+        Logger.recordOutput("Tracer/" + traceTime.getKey(), traceTime.getValue());
+        entryArray.add(traceTime.getKey());
+      }
+
+      // log gc time
+      if (gcs.size() > 0) Logger.recordOutput("Tracer/GCTime", gcTimeThisCycle);
+      gcTimeThisCycle = 0.0;
+
+      // clean up state
+      traceTimes.clear();
+      traceStackHistory.clear();
+    }
+  }
+
+  private static final ThreadLocal<TracerState> threadLocalState =
+      ThreadLocal.withInitial(
+          () -> {
+            return new TracerState(Thread.currentThread().getName());
+          });
+
+  public static void disableGcLoggingForCurrentThread() {
+    TracerState state = threadLocalState.get();
+    state.gcs.clear();
+  }
+
+  private static void startTrace(final String name, final TracerState state) {
+    String stack = state.appendTraceStack(name);
+    TraceStartData data = state.traceStartTimes.get(stack);
+    if (data == null) {
+      data = new TraceStartData();
+      state.traceStartTimes.put(stack, data);
+    }
+    data.set(Logger.getRealTimestamp() / 1000.0, state.totalGCTime());
+  }
+
+  private static void endTrace(final TracerState state) {
+    try {
+      String stack = state.popTraceStack();
+      var startData = state.traceStartTimes.get(stack);
+      double gcTimeSinceStart = state.totalGCTime() - startData.startGCTotalTime;
+      state.gcTimeThisCycle += gcTimeSinceStart;
+      state.traceTimes.put(
+          stack, Logger.getRealTimestamp() / 1000.0 - startData.startTime - gcTimeSinceStart);
+      if (state.traceStack.size() == 0) {
+        state.endCycle();
+      }
+    } catch (Exception e) {
+      DriverStation.reportError(
+          "[Tracer] An end trace was called with no opening trace " + e, true);
+    }
+  }
+
+  /**
+   * Starts a trace, should be called at the beginning of a function thats not being called by user
+   * code. Should be paired with {@link Tracer#endTrace()} at the end of the function.
+   *
+   * <p>Best used in periodic functions in Subsystems and Robot.java.
+   *
+   * @param name the name of the trace, should be unique to the function.
+   */
+  public static void startTrace(String name) {
```

Comment:
I'm not sure why this or the endTrace variant should exist at all. You should just use traceFunc.